HDDS-11463. Track and display failed DataNode storage locations in SCM.

slfan1989 opened the pull request:
## What changes were proposed in this pull request?

Currently, we lack tools on the SCM side to track failed disks on DataNodes. DataNodes have already reported this information, and we need to display it.

In this PR, we will display the failed disks on the DataNode. The information can be displayed in JSON format or using the default format.

> Default format

```
Datanode Volume Failures (5 Volumes)

Node         : localhost-62.238.104.185 (de97aaf3-99ad-449d-ad92-2c4f5a744b49) 
Failed Volume: /data0/ozonedata/hdds 
Capacity Lost: 7430477791683 B (6.76 TB) 
Failure Date : Thu Oct 03 09:25:16 +0800 2024 

Node         : localhost-163.120.165.68 (cf40e987-8952-4f7a-88b7-096e6b285243) 
Failed Volume: /data1/ozonedata/hdds 
Capacity Lost: 7430477791683 B (6.76 TB) 
Failure Date : Thu Oct 03 09:25:16 +0800 2024 

Node         : localhost-253.243.206.120 (0cc77921-489d-4cf0-a036-475faa16d443) 
Failed Volume: /data2/ozonedata/hdds 
Capacity Lost: 7430477791683 B (6.76 TB) 
Failure Date : Thu Oct 03 09:25:16 +0800 2024 

Node         : localhost-136.194.243.81 (5cb6430d-0ce5-4204-b265-179ee38fb30e) 
Failed Volume: /data3/ozonedata/hdds 
Capacity Lost: 7430477791683 B (6.76 TB) 
Failure Date : Thu Oct 03 09:25:16 +0800 2024 

Node         : localhost-48.253.209.226 (f99a8374-edb0-419d-9cba-cfab9d9e8a2e) 
Failed Volume: /data4/ozonedata/hdds 
Capacity Lost: 7430477791683 B (6.76 TB) 
Failure Date : Thu Oct 03 09:25:16 +0800 2024 
```

> Json format

```
[ {
  "node" : "localhost-161.170.151.131 (155bb574-7ed8-41cd-a868-815f4c2b0d60)",
  "volumeName" : "/data0/ozonedata/hdds",
  "failureDate" : 1727918794694,
  "capacityLost" : 7430477791683
}, {
  "node" : "localhost-67.218.46.23 (520d29eb-8387-4cda-bcb1-8727fdddd451)",
  "volumeName" : "/data1/ozonedata/hdds",
  "failureDate" : 1727918794695,
  "capacityLost" : 7430477791683
}, {
  "node" : "localhost-30.151.88.21 (d66cab50-bbf8-4199-9d7f-82da84a30137)",
  "volumeName" : "/data2/ozonedata/hdds",
  "failureDate" : 1727918794695,
  "capacityLost" : 7430477791683
}, {
  "node" : "localhost-78.50.38.217 (a673f50a-6f74-4e62-8c0c-f7337d5f3ce5)",
  "volumeName" : "/data3/ozonedata/hdds",
  "failureDate" : 1727918794695,
  "capacityLost" : 7430477791683
}, {
  "node" : "localhost-138.205.52.25 (84b7e49a-9bd4-4115-96fa-69f2d259343c)",
  "volumeName" : "/data4/ozonedata/hdds",
  "failureDate" : 1727918794695,
  "capacityLost" : 7430477791683
} ]
```

> Table format

```
+-------------------------------------------------------------------------------------------------------------------------------------------+
|                                                         Datanode Volume Failures                                                          |
+------------------------------------------------------------------+-----------------------+---------------+--------------------------------+
|                               Node                               |      Volume Name      | Capacity Lost |          Failure Date          |
+------------------------------------------------------------------+-----------------------+---------------+--------------------------------+
|  localhost-83.212.219.28 (8b6addb1-759a-49e9-99fb-0d1a6cfb2d7f)  | /data0/ozonedata/hdds |    6.76 TB    | Sat Oct 05 17:47:47 +0800 2024 |
| localhost-103.199.236.47 (0dbe503a-3382-4753-b95a-447bab5766c4)  | /data1/ozonedata/hdds |    6.76 TB    | Sat Oct 05 17:47:47 +0800 2024 |
|  localhost-178.123.46.32 (2017076a-e763-4f47-abce-78535b5770a3)  | /data2/ozonedata/hdds |    6.76 TB    | Sat Oct 05 17:47:47 +0800 2024 |
| localhost-123.112.235.228 (aaebb6a7-6b62-4160-9934-b16b8fdde65e) | /data3/ozonedata/hdds |    6.76 TB    | Sat Oct 05 17:47:47 +0800 2024 |
| localhost-249.235.216.19 (cbc7c0b5-5ae0-4e40-91b8-1d9c419a007c)  | /data4/ozonedata/hdds |    6.76 TB    | Sat Oct 05 17:47:47 +0800 2024 |
+------------------------------------------------------------------+-----------------------+---------------+--------------------------------+
```

## What is the link to the Apache JIRA

JIRA: HDDS-11463. Track and display failed DataNode storage locations in SCM.

## How was this patch tested?

(Please explain how this patch was tested. Ex: unit tests, manual tests, workflow run on the fork git repo.)
(If this patch involves UI changes, please attach a screenshot; otherwise, remove this.)

slfan1989 commented:
@errose28 Could you please help review this PR? Thank you very much! We discussed the relevant implementation together in HDDS-11463.

errose28 commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/admin/scm/VolumeFailureSubCommand.java:
```diff
@@ -0,0 +1,101 @@
+ * Handler of ozone admin scm volumesfailure command.
+ */
+@Command(
+    name = "volumesfailure",
```
For the CLI, we should probably use something like `ozone admin datanode volume list`. The `datanode` subcommand is already used to retrieve information about datanodes from SCM. Splitting the commands so that `volume` has its own subcommand gives us more options in the future.

To distinguish failed and healthy volumes and filter out different nodes, we can either add some kind of filter flag, or leave it up to grep/jq to be applied to the output.

errose28 replied:
This also means we should make the RPC more generic to support pulling all volume information.

slfan1989 replied:
Thank you for helping to review this PR! I will continue to improve the relevant code based on your suggestions.

errose28 commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/volume/StorageVolume.java:
```diff
@@ -382,6 +383,7 @@ public abstract static class Builder<T extends Builder<T>> {
     private boolean failedVolume = false;
     private String datanodeUuid;
     private String clusterID;
+    private long failureDate;
```
Lets use `failureTime`. I'm assuming this is being stored as millis since epoch, so it will have data and time information.

slfan1989 replied:
I have improved the relevant code.

errose28 commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/volume/StorageVolume.java:
```diff
@@ -548,6 +571,11 @@ public ConfigurationSource getConf() {
+    // Ensure it is set only once,
+    // which is the time when the failure was first detected.
+    if (failureDate == 0L) {
+      setFailureDate(Time.now());
```
Let's use `Instant.now()` per HDDS-7911.

slfan1989 replied:
@errose28 Can you help review this PR again? Thank you very much!

errose28 reviewed (commented):
Thanks for working on this @slfan1989, this looks like a useful addition. I only had time for a quick high level look for now.

adoroszlai commented:
Thanks @slfan1989 for working on this.  Converted it to draft because there is a failing test:

```
[ERROR] org.apache.hadoop.hdds.scm.node.TestSCMNodeManager.tesVolumeInfoFromNodeReport  Time elapsed: 1.105 s  <<< ERROR!
java.lang.UnsupportedOperationException
	at java.base/java.util.AbstractList.add(AbstractList.java:153)
	at java.base/java.util.AbstractList.add(AbstractList.java:111)
	at org.apache.hadoop.hdds.scm.node.DatanodeInfo.updateStorageReports(DatanodeInfo.java:186)
	at org.apache.hadoop.hdds.scm.node.SCMNodeManager.processNodeReport(SCMNodeManager.java:674)
	at org.apache.hadoop.hdds.scm.node.SCMNodeManager.register(SCMNodeManager.java:423)
	at org.apache.hadoop.hdds.scm.node.SCMNodeManager.register(SCMNodeManager.java:360)
	at org.apache.hadoop.hdds.scm.node.TestSCMNodeManager.tesVolumeInfoFromNodeReport(TestSCMNodeManager.java:1591)
```

https://github.com/slfan1989/ozone/actions/runs/11471452180
https://github.com/slfan1989/ozone/actions/runs/11476535807
https://github.com/slfan1989/ozone/actions/runs/11625983369

slfan1989 commented:
> Thanks @slfan1989 for working on this. Converted it to draft because there is a failing test:
> 
> ```
> [ERROR] org.apache.hadoop.hdds.scm.node.TestSCMNodeManager.tesVolumeInfoFromNodeReport  Time elapsed: 1.105 s  <<< ERROR!
> java.lang.UnsupportedOperationException
> 	at java.base/java.util.AbstractList.add(AbstractList.java:153)
> 	at java.base/java.util.AbstractList.add(AbstractList.java:111)
> 	at org.apache.hadoop.hdds.scm.node.DatanodeInfo.updateStorageReports(DatanodeInfo.java:186)
> 	at org.apache.hadoop.hdds.scm.node.SCMNodeManager.processNodeReport(SCMNodeManager.java:674)
> 	at org.apache.hadoop.hdds.scm.node.SCMNodeManager.register(SCMNodeManager.java:423)
> 	at org.apache.hadoop.hdds.scm.node.SCMNodeManager.register(SCMNodeManager.java:360)
> 	at org.apache.hadoop.hdds.scm.node.TestSCMNodeManager.tesVolumeInfoFromNodeReport(TestSCMNodeManager.java:1591)
> ```
> 
> https://github.com/slfan1989/ozone/actions/runs/11471452180 https://github.com/slfan1989/ozone/actions/runs/11476535807 https://github.com/slfan1989/ozone/actions/runs/11625983369

@adoroszlai Thank you for reviewing this PR！ I am currently making improvements, and once the changes pass the CI tests in my branch, I will reopen the PR.

cc: @errose28

slfan1989 commented:
@adoroszlai Thank you for reviewing this PR! I will also pay closer attention to CI issues in future development. I understand that CI testing resources are valuable.  

I have made improvements to the code based on @errose28 suggestions and also fixed the related unit test errors. The CI for my branch has passed(https://github.com/slfan1989/ozone/actions/runs/11719380711), and I have updated the PR status to 'Ready for Review'.

slfan1989 commented:
@errose28 Could you please help review this PR again? Thank you very much! I’ve made some additional improvements to this PR, as we wanted to print all the disk information. However, since there’s quite a lot of disk data, I’ve added pagination functionality.

adoroszlai commented:
Temporarily converted to draft and assigned to myself, to resolve conflicts.

slfan1989 commented:
@adoroszlai Thank you for your attention to this PR. I will continue to follow up on it.

adoroszlai commented:
Merged from `master`.  There will be one checkstyle problem:

```
hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/scm/datanode/package-info.java
 18: Missing a Javadoc comment.
 18: Missing javadoc for package-info.java file.
```

Previously the license header was a javadoc in this new file, so the problem was hidden.

adoroszlai commented:
```
Javadoc Warnings
hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/protocol/StorageContainerLocationProtocolServerSideTranslatorPB.java:1365: warning - @param argument "displayMode" is not a parameter name.
hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/protocol/StorageContainerLocationProtocolServerSideTranslatorPB.java:1368: warning - @param argument "uuid" is not a parameter name.
hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/protocol/StorageContainerLocationProtocolServerSideTranslatorPB.java:1369: warning - @param argument "hostName" is not a parameter name.
hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/protocol/StorageContainerLocationProtocolServerSideTranslatorPB.java:1370: warning - @param argument "pageSize" is not a parameter name.
hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/protocol/StorageContainerLocationProtocolServerSideTranslatorPB.java:1371: warning - @param argument "currentPage" is not a parameter name.
```

https://github.com/apache/ozone/actions/runs/13468894567/job/37640031573#step:8:8432

slfan1989 commented:
@adoroszlai Could you help review this PR? Thank you very much! It would be a useful operation to get the faulty disk from SCM. 
cc: @errose28

slfan1989 commented:
@xichen01 Could you please help review this PR? Thank you very much!

adoroszlai commented on hadoop-ozone/cli-admin/src/main/java/org/apache/hadoop/hdds/scm/cli/datanode/VolumeSubCommand.java:
```diff
@@ -0,0 +1,222 @@
+       description = "Format output as Table.")
+  private boolean table;
+
+  // PageSize refers to the number of items displayed per page
+  // in a paginated view.
+  @Option(names = { "--pageSize" },
+       defaultValue = "20",
+       description = "The number of volume information items displayed per page.")
+  private int pageSize;
+
+  // The current page.
+  @Option(names = { "--currentPage" },
+       defaultValue = "1",
+       description = "The current page.")
+  private int currentPage;
```
Other Ozone CLI commands allow pagination (via `ListOptions` reusable mixin) with `--start` to specify the start item, `--length` for "page size", and `--all` (only one of these last two are allowed).  Please try to use options consistent with that.  Using `ListOption` directly may not work, since `--prefix` does not seem to be applicable here.

slfan1989 replied:
> Other Ozone CLI commands allow pagination (via ListOptions reusable mixin) with --start to specify the start item, --length for "page size", and --all (only one of these last two are allowed). Please try to use options consistent with that. 

I have thoroughly reviewed `ListOptions`, and it is indeed a powerful tool class. However, my requirements differ slightly from its original functionality. I want to implement a pagination display, as there may sometimes be a certain number of damaged disks. For example, if there are 30 damaged disks and the `limit` is set to 10, the `currentPage` can range from 1 to 3, allowing for pagination. In this case, `currentPage` represents the current page number. Therefore, I plan to add the `currentPage` option to `ListOptions` to meet this need.

> Using ListOption directly may not work, since --prefix does not seem to be applicable here.

The `--start` and `--prefix` options have limited relevance for our functionality. Currently, we support filtering by `hostName`, which I believe should be sufficient. I'm unsure whether we need to filter by disk letter as well. What do you think?

adoroszlai replied:
The problem with numeric "current page" is that the list is not fixed.  By the time you request the next page, it may be shorter or longer.  Then you may get the same item twice (if new item is added in the range of the earlier pages) or not at all (if item is removed from earlier pages).

Using an anchor item eliminates the problem of duplicates.

Compare GitHub's list of commits, which uses commit SHA as anchor:

https://github.com/apache/ozone/commits/master?after=e6daae4af3f9839be52ea6db636c9ff434acf306+34

to list of pull requests, which has simple numeric page:

https://github.com/apache/ozone/pulls?page=2

Created HDDS-12995 to move out `--prefix`, to allow reusing `ListOptions` for lists that are filtered by other parameters.

slfan1989 replied:
Thank you for your detailed explanation! I now understand the reliability of using anchor points and plan to make the corresponding improvements in the code. Regarding HDDS-12995, I understand that you mean to move the `--prefix` option from its current position, specifically placing it within the command, rather than having it as a global option. Am I understanding this correctly?

adoroszlai replied:
Yes.

adoroszlai replied:
#8413

adoroszlai commented on hadoop-ozone/cli-admin/src/main/java/org/apache/hadoop/hdds/scm/cli/datanode/VolumeSubCommand.java:
```diff
@@ -0,0 +1,222 @@
+    versionProvider = HddsVersionProvider.class)
+public class VolumeSubCommand extends ScmSubcommand {
+
+  // Display it in JSON format.
+  @Option(names = { "--json" },
+       defaultValue = "false",
+       description = "Format output as JSON.")
+  private boolean json;
+
+  // Display it in TABLE format.
+  @Option(names = { "--table" },
+       defaultValue = "false",
+       description = "Format output as Table.")
+  private boolean table;
```
Please make these [exclusive](https://picocli.info/#_mutually_exclusive_options), like:

https://github.com/apache/ozone/blob/af1f98cb3737671df6dce0bdd4af07f8b510dfe7/hadoop-ozone/cli-shell/src/main/java/org/apache/hadoop/ozone/shell/ListOptions.java#L27-L28

https://github.com/apache/ozone/blob/af1f98cb3737671df6dce0bdd4af07f8b510dfe7/hadoop-ozone/cli-shell/src/main/java/org/apache/hadoop/ozone/shell/ListOptions.java#L63-L74

slfan1989 replied:
That's a good suggestion! I will improve the code based on this recommendation.

adoroszlai commented on hadoop-ozone/cli-admin/src/main/java/org/apache/hadoop/hdds/scm/cli/datanode/VolumeSubCommand.java:
```diff
@@ -0,0 +1,222 @@
+  // The UUID identifier of the DataNode.
+  @Option(names = { "--uuid" },
+      defaultValue = "",
+      description = "failed is used to display failed disks, " +
+      "normal is used to display normal disks.")
```
description seems to be dup of `--displayMode`.  (Same for `--hostName`.)

slfan1989 replied:
I will improve the help information for `displayMode`, `hostname`, and `uuid`.

adoroszlai commented on hadoop-ozone/cli-admin/src/main/java/org/apache/hadoop/hdds/scm/cli/datanode/VolumeSubCommand.java:
```diff
@@ -0,0 +1,222 @@
+   * 'failed' is used to display failed disks,
+   * and 'normal' is used to display normal disks.
+   */
+  @Option(names = { "--displayMode" },
+      defaultValue = "all",
+      description = "failed is used to display failed disks, " +
+      "normal is used to display normal disks.")
+  private String displayMode;
```
Please use an `enum` to limit values allowed.

https://picocli.info/#_enum_types

adoroszlai commented on hadoop-ozone/cli-admin/src/main/java/org/apache/hadoop/hdds/scm/cli/datanode/VolumeSubCommand.java:
```diff
@@ -0,0 +1,222 @@
+
+    // Print the relevant information if the return value is empty.
+    if (response == null || CollectionUtils.isEmpty(response.getVolumeInfosList())) {
+      System.out.println("No volume data was retrieved.");
```
I'm not sure this informational message is needed, but if you think so, at least move it after handling JSON format, which should output valid JSON even for empty results.

adoroszlai commented on hadoop-ozone/cli-admin/src/main/java/org/apache/hadoop/hdds/scm/cli/datanode/VolumeSubCommand.java:
```diff
@@ -0,0 +1,222 @@
+
+    // If displayed in JSON format.
+    if (json) {
+      System.out.print(JsonUtils.toJsonStringWithDefaultPrettyPrinter(volumeInfos));
```
Please use `out()` and `err()` inherited from `AbstractSubcommand` for output.

slfan1989 replied:
This is a good suggestion and I have improved this part of the code.

adoroszlai commented on hadoop-ozone/cli-admin/src/test/java/org/apache/hadoop/ozone/scm/TestVolumeCommand.java:
```diff
@@ -0,0 +1,164 @@
+  @BeforeEach
+  public void setup() throws UnsupportedEncodingException {
+    cmd = new VolumeSubCommand();
+    System.setOut(new PrintStream(outContent, false, DEFAULT_ENCODING));
+    System.setErr(new PrintStream(errContent, false, DEFAULT_ENCODING));
```
Do we need to manually set out/err when using `SystemOutCapturer`?

slfan1989 replied:
This is a good question! After using `SystemOutCapturer`, we no longer need to manually set out/err. I have already improved this part of the code.

adoroszlai commented on hadoop-ozone/cli-admin/src/test/java/org/apache/hadoop/ozone/scm/TestVolumeCommand.java:
```diff
@@ -0,0 +1,164 @@
+    return builder.build();
+  }
+
+  private HddsProtos.DatanodeDetailsProto createDatanodeDetails() {
+    Random random = ThreadLocalRandom.current();
+    String ipAddress = random.nextInt(256)
+        + "." + random.nextInt(256)
+        + "." + random.nextInt(256)
+        + "." + random.nextInt(256);
+
+    DatanodeDetails.Builder dn = DatanodeDetails.newBuilder()
+        .setUuid(UUID.randomUUID())
+        .setHostName("localhost" + "-" + ipAddress)
+        .setIpAddress(ipAddress)
+        .setPersistedOpState(HddsProtos.NodeOperationalState.IN_SERVICE)
+        .setPersistedOpStateExpiry(0);
+
+    for (DatanodeDetails.Port.Name name : ALL_PORTS) {
+      dn.addPort(DatanodeDetails.newPort(name, 0));
+    }
```
Use `MockDatanodeDetails`.

adoroszlai commented on hadoop-hdds/interface-client/src/main/proto/hdds.proto:
```diff
@@ -304,6 +304,15 @@ message RemoveScmResponseProto {
 }
 
+message VolumeInfoProto {
+    required string uuid = 1;
+    required string hostName = 2;
+    required string volumeName = 3;
+    required bool failed = 4;
+    required int64 failureTime = 5;
+    required int64 capacity = 6;
```
Add fields as `optional`.

https://protobuf.dev/programming-guides/proto2/#field-labels

adoroszlai commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/DatanodeInfo.java:
```diff
@@ -49,7 +51,7 @@ public class DatanodeInfo extends DatanodeDetails {
   private long lastStatsUpdatedTime;
   private int failedVolumeCount;
-
+  private List<VolumeInfoProto> volumeInfos;
```
`volumeInfos` is never assigned.  I think the intention was to update this variable in `updateStorageReports` in the block holding `writeLock`.

slfan1989 replied:
Thank you for pointing out this issue! I hadn't noticed these details before.

adoroszlai commented on hadoop-ozone/cli-admin/src/main/java/org/apache/hadoop/hdds/scm/cli/datanode/VolumeSubCommand.java:
```diff
@@ -0,0 +1,222 @@
+ * We provide a set of volume commands to display the disk information of the DataNode.
+ */
+@Command(
+    name = "volumes",
+    description = "Display the list of volumes on the DataNode.",
+    mixinStandardHelpOptions = true,
+    versionProvider = HddsVersionProvider.class)
+public class VolumeSubCommand extends ScmSubcommand {
```
Currently this subcommand is dangling, not added to any parent.  If you add it somewhere, it will be available as `ozone ... volumes`.  [Suggestion](https://github.com/apache/ozone/pull/7266#discussion_r1819736586) from @errose28 was to add this as `ozone admin datanode volume list`, which allows adding other subcommands related to datanode volumes later.

Subcommands are added like this:

https://github.com/apache/ozone/blob/af1f98cb3737671df6dce0bdd4af07f8b510dfe7/hadoop-ozone/cli-admin/src/main/java/org/apache/hadoop/hdds/scm/cli/datanode/DatanodeCommands.java#L33-L40

adoroszlai reviewed (changes requested):
Thanks @slfan1989 for the patch.

slfan1989 commented:
@adoroszlai Thank you very much for reviewing this PR! I will continue to improve it.

slfan1989 commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/SCMClientProtocolServer.java:
```diff
@@ -1633,4 +1639,89 @@ public String getMetrics(String query) throws IOException {
+          break;
+        }
+      }
+    }
```
@adoroszlai I added logic to skip startItem in this part of the code, but after thinking it through, I realized it’s better to use the server’s hostname or UUID as startItem instead of a disk prefix. That’s because many machines name their disks like data0 to data9, and using a disk name could lead to unexpected filtering behavior.

slfan1989 commented:
@adoroszlai Could you help review this PR? Thank you very much! 

I have conducted tests on my own branch, and it currently passes the key CI tests.

https://github.com/slfan1989/ozone/actions/runs/15209585380

errose28 commented:
Hi @slfan1989 thanks for working on this change. I think there are three attributes being added here which should be reviewed separately:
1. Adding an SCM RPC to retrieve volume information
2. Tracking failure time of the volume
3. Adding a CLI to view the volume information

The RPC to retrieve volume information is definitely required going forward regardless of the other two items to create some sort of CLI to query volume state.

Tracking the failure time of the volume seems like a somewhat invasive change since it spans the datanode, heartbeat, and SCM. Is this necessary, or is it enough to depend on a metrics database to track timing of cluster events? Of course we need improvements to our volume metrics as well as mentioned in #8405.

On the CLI front, I do think we need a dedicated `ozone admin datanode info` command going forward as outlined in HDDS-13097. This would give all volume information per node. With volume counters added to `ozone admin datanode list` as proposed in HDDS-13096, we could get all failed volumes in a two step process:
1. jq filter on `ozone admin datanode list` to find all nodes with failed volumes.
2. jq filter on `ozone admin datanode info` to get specific information about the failed volumes, including their capacity.

Do we need a dedicated `ozone admin datanode volume list/info` command pairing in addition to this? It may be useful to have such cross-cutting commands to get information in one shot, but on the other hand it may result in duplication at the CLI. For example I could see the request to add node filtering to `ozone admin datanode volume list/info` at which point it becomes much the same as `ozone admin datanode list/info`.

slfan1989 commented:
@errose28 Thank you for your message! I'd like to share some thoughts from a different perspective. As it stands, this feature does not conflict with the proposal in #8405. #8405 represents a more innovative and forward-looking design, and although it's still under discussion, it will certainly be valuable if implemented as planned.

At the same time, I believe this feature does not impact HDDS-13096 or HDDS-13097. My comment on #8405 was more about expressing expectations for the system’s future capabilities — I hope Ozone can gradually support such features — rather than raising any objections to #8405 itself.

The design of #7266 is inspired by HDFS's disk failure detection mechanism, with the goal of improving the system's ability to identify and locate failed disks. For users migrating from HDFS to Ozone, using the volume command to directly view failed disks can offer a more intuitive and convenient operational experience.

From my perspective, we all play different roles in this project. Your team focuses on evolving and optimizing the system's architecture, while we, as external users, are more focused on refining specific functional details based on real-world use. Ultimately, however, we share the same goal: to make Ozone more robust, more user-friendly, and more widely adopted.

Naturally, it's not easy to fully align these detail-oriented changes with larger, ongoing feature developments — for example, making #7266 fully consistent with #8405. This is mainly because #8405 is broader in scope, with a longer timeline, whereas #7266 focuses on a very specific aspect. While we fully respect the overall direction, we also hope to move forward with some smaller, incremental improvements to address current practical issues.

In addition to this PR, we're also working on several other enhancements. For instance, we've implemented mechanisms to collect DataNode I/O statistics to more precisely manage container replication. We've also introduced time-based peak/off-peak control logic for various DataNode management operations (such as deletion, container replication, and EC container reconstruction). These improvements are driven by real-world production needs, and from our perspective, they've shown positive results.

However, since many of these PRs have some degree of code coupling with our previous contributions, it's difficult for us to combine everything into a single, unified patch for upstream submission.

Therefore, we hope to proceed with #7266 for now. If #8405 later results in a more complete or improved solution, we’d be happy to continue refining things in that direction. In the meantime, this also gives us a valuable opportunity to participate in the community and contribute to Ozone’s development.

slfan1989 commented:
@adoroszlai Thank you very much for your detailed suggestions! I've made the changes accordingly. Could you review this PR again? Thank you very much!  I respect @errose28's perspective. However, I believe this PR does not conflict with #8405, nor with HDDS-13096 or HDDS-13097 — they can coexist. We've already spent considerable time reviewing this PR together, and I'd like to continue moving it forward.

adoroszlai commented:
@slfan1989 Thanks for all your efforts on this PR.  The concerns/suggestions raised by @errose28 make sense though.  Please try to reach agreement.

I won't be able to re-review until next week in any case.

slfan1989 commented:
> @slfan1989 Thanks for all your efforts on this PR. The concerns/suggestions raised by @errose28 make sense though. Please try to reach agreement.
> 
> I won't be able to re-review until next week in any case.

@adoroszlai Thank you very much for your message and for your continued support and assistance! Since @errose28 is currently planning some new features, I believe this PR could be considered as part of that effort, especially given the amount of work we've already invested. As for which specific features should be retained, it would be helpful if @errose28 could review and provide guidance.

errose28 commented:
Hi @slfan1989 I appreciate your response and the work you've done on these changes. My suggestion to split this PR down was not meant to diminish any of the work that has been put in here, but to speed up in incorporating the work into Ozone.

A change like this approaching 1k lines, 70+ review comments,  and encompassing multiple items is going to be large for any reviewer, and I think we could make faster progress by splitting it. For example, I think we could iterate on the volume info RPC in SCM pretty quickly and get that change merged first. I'm ok with adding volume failure time and capacity lost to SCM as well, but it will be easier to review those as their own change. This way most of the work here can be merged while we discuss the CLI.

> The design of [#7266](https://github.com/apache/ozone/pull/7266) is inspired by HDFS's disk failure detection mechanism, with the goal of improving the system's ability to identify and locate failed disks. For users migrating from HDFS to Ozone, using the volume command to directly view failed disks can offer a more intuitive and convenient operational experience.

Can you add more details about how this is similar to HDFS? I'm familiar with `hdfs dfsadmin -report` to list failed volumes, but that breaks the information down at the node level with failed volume counters for each node, which seems more similar to an `ozone admin datanode list` command.

> I believe this feature does not impact [HDDS-13096](https://issues.apache.org/jira/browse/HDDS-13096) or [HDDS-13097](https://issues.apache.org/jira/browse/HDDS-13097).

Yes we could have both `ozone admin datanode list/info` and `ozone admin datanode volume list` without code conflicts, but we need to build a maintainable and intuitive CLI, which means we should avoid commands that do the same or similar things. In this case I think we should standardize one way to get volume information from the CLI. I propose keeping this within the `datanode info` command because volumes are completely contained within a node, unlike cross cutting concepts like containers and pipelines which have their own subcommands. Based on early comments like [this](https://github.com/apache/ozone/pull/8405#discussion_r2106224868) I think we would end up needing node filtering in `ozone admin datanode volume list` at which point it becomes very similar to `ozone admin datanode list/info`.

slfan1989 commented:
@errose28 @adoroszlai I have completed the improvements for this PR and kept the RPC interface part. Could you please help review it? Thank you very much!

slfan1989 commented:
> Sorry for the delay. Can we have this PR be just the SCM <-> client communication for querying volume info? Right now it also contains information for SCM <-> DN communication about the failure time of the volume, which is not directly related and can be added in a follow-up change. The new RPC will also need tests added.

@errose28 Thank you for your feedback. I will improve this PR based on your suggestions.
